### ISSUE TITLE
Slow incremental webpack build times

### ISSUE THREAD

**Issue by trotzig** [Updater/node-sass-json-importer#15]
I understand that this might not be the right place to report this issue, but I thought I'd post it here first to see if anyone else has more context. 

We have a fairly large React app where we use webpack to compile our application.js bundle. Stylesheets are resolved using sass-loader, and components import their own stylesheets, like this:

``` jsx
const styles = require('./index.scss');

const Component = React.createClass({
  render() {
    return <div className={styles.
  }
});
```

To share configuration between the components and sass files (we use scss), we have a `shared_config.json` file with some key-value settings. In the past, we've used a manual script to generate a `shared_config.scss` file that we then use in our scss files. Yesterday, (thanks to https://github.com/Updater/node-sass-json-importer/pull/13) we switched to directly importing the json file, through the help of `node-sass-json-import`. That worked out well, the only problem is that our incremental build times for our webpack bundle went up from ~2.5 seconds to ~12 seconds. An incremental build happens any time a `.scss` or `.js*` file is saved. I've spent some time to bisect our git history to make sure that it was in fact adding this to our `webpack.config.js` that introduced the performance regression: 

``` js
  sassLoader: {
    importer: require('node-sass-json-importer'),
    includePaths: [
      path.resolve('./app/assets/stylesheets'),
    ],
  }
```

Any ideas what could be the problem? I'm suspecting that node-sass/libsass isn't being smart enough with custom importers (after all, [node-sass warns about this being an experimental feature](https://github.com/sass/node-sass#importer--v200---experimental)).

**Comment by pmowrer** [comment]
That's bizarre, @trotzig. Haven't experienced any performance issues ourselves. Please let us know if you discover anything that might help us identify the problem.

**Comment by trotzig** [comment]
Will do, thought there's a chance I won't have time to look at this any time soon.

**Comment by pmowrer** [comment]
Closing until we know of a way to reproduce.

**Comment by trotzig** [comment]
@pmowrer I'm sorry I couldn't provide more information about this.

**Comment by harlanlewis** [comment]
Hesitant to revive an old issue, but came across the issue exactly as described above.

Webpack boot time was unaffected, however hot reload for single-line JSX whitespace changes increased 10x from ~0.9s to ~9s. I've isolated this package as the source of the increase, as uninstalling and defining our variables directly in Sass is the only change between the build times.

- webpack 1.12.12
- react 0.14.7
- node-sass 3.4.2

Is there anything I can do to help reproduce this issue? We're quite keen on the approach!

**Comment by pmowrer** [comment]
@harlanlewis Thanks for the report! 

I'd be interested to know how/where those ~8s accumulate.

Maybe to start, clone the repo and `npm link` it to the project that depends on `node-sass-json-importer`. Then put some `console.log` statements in the source, perhaps a line at the [top](https://github.com/Updater/node-sass-json-importer/blob/master/src/index.js#L6) to get a sense of how many files are processed and what "time" it is when each file starts processing. Maybe that would reveal a certain import taking a very long time to process.

**Comment by harlanlewis** [comment]
Thanks for the pointers to debug! I'll look into it within the next few days and report back.

**Comment by pmowrer** [comment]
Sounds good! Let me know if you need any further instructions... I didn't go into detail about how `npm link` works and also you'll need to run `npm run compile` within the cloned repo any time you make changes to the source to produce a new bundle.

**Comment by JaZo** [comment]
@harlanlewis I'm experiencing this too (~0.2s to ~12s). Do you have an update on your debug sessions?

I'm using the following versions:
- webpack 2.4.1
- node-sass 4.5.2
- node-sass-json-importer 3.0.2

**Comment by harlanlewis** [comment]
I'm afraid not, this slipped off the priority list.

**Comment by samtgarson** [comment]
I'm also seeing this issue, using webpack 1.12.6 and node-sass 3.4.2. My build times have gone from sub 2 seconds to between 20-30 seconds every time... Making this kinda unusable :/

Don't have time to delve into this further at the moment but would be good to hear if anyone makes any progress.

**Comment by pmowrer** [comment]
FWIW, we've been using it with both webpack 1 and 2, never experiencing a problem. We have a pretty limited amount of JSON that's getting imported though. It'd be interesting to see if this problem is due to some specific JSON construct.

**Comment by samtgarson** [comment]
My JSON was a one-level object with 6 keys with all numeric values... but I have a large codebase with the file getting imported many times 🤔

**Comment by pmowrer** [comment]
That is true of our applications as well. Large code bases where constants are getting imported into many files.

**Comment by harlanlewis** [comment]
Ours was a bit different:
- 5-10 json files with 1 object per file, depth <3 in in all objects, 
- each json file imported just once into a single sass file per json (which provided access to values via functions etc),
- those sass files imported many times throughout the app.

Sorry I don't have the bandwidth at present to do the more thorough performance profile @pmowrer recommended back in February!

**Comment by JaZo** [comment]
I've done some profiling as @pmowrer suggested and put some logs at the [beginning](https://github.com/Updater/node-sass-json-importer/blob/master/src/index.js#L6) and [end](https://github.com/Updater/node-sass-json-importer/blob/master/src/index.js#L43) of the importer. The timing difference is only ~5ms! These results do not seem to indicate that the speed decrease is caused by node-sass-json-importer itself, but rather by something deeper in the stack. I'll try to see if I can dig a little deeper.

N.B. My json is very simple:
`{"screen-xs": "480px", "screen-sm": "768px", "screen-m": "900px", "screen-l": "1100px", "screen-xl": "1300px", "screen-xxl": "1600px", "screen-mega": "1900px"}`

**Comment by JaZo** [comment]
Minor update! I was using Laravel Mix to build my assets, but it had an issue which was causing my speed decrease: https://github.com/JeffreyWay/laravel-mix/issues/232. We are several months later since my last tests/profiling and I ditched Laravel Mix in favor of 'vanilla' Webpack. Today, I decided to give this importer another try. The speed decrease is now much smaller: ~0.2s to ~0.8s. I'd say the increase of ~0.6s is acceptable!

N.B. I'm currently using the following versions:
- webpack 3.6.0
- node-sass 4.5.3
- node-sass-json-importer 3.0.2
